cartography/intel/gcp/gcf.py

```diff
@@ -0,0 +1,127 @@
+from cartography.graph.job import GraphJob
+from cartography.models.gcp.gcf import GCPCloudFunctionSchema
+from cartography.util import timeit
+
+logger = logging.getLogger(__name__)
+
+
+@timeit
+def get_gcp_cloud_functions(project_id: str, functions_client: Resource) -> List[Dict]:
+    """
+    Fetches raw GCP Cloud Functions data for a given project.
+    """
+    logger.info(f"Collecting Cloud Functions for project: {project_id}")
+    collected_functions = []
+    try:
+        parent = f"projects/{project_id}/locations/-"
+        request = functions_client.projects().locations().functions().list(parent=parent)
+        while request is not None:
+            response = request.execute()
+            if 'functions' in response:
+                collected_functions.extend(response['functions'])
+            request = functions_client.projects().locations().functions().list_next(
+                previous_request=request,
+                previous_response=response,
+            )
+        return collected_functions
+    except HttpError as e:
+        error_json = json.loads(e.content.decode("utf-8"))
+        err = error_json.get("error", {})
+        if (
+            err.get("status", "") == "PERMISSION_DENIED"
+            or (err.get("message") and "API has not been used" in err.get("message"))
+        ):
+            logger.warning(
+                (
+                    "Could not retrieve Cloud Functions on project %s due to permissions issues or API not enabled. Code: %s, Message: %s"
+                ),
+                project_id,
+                err.get("code"),
+                err.get("message"),
+            )
+            return []
+        else:
+            raise
+
+
+@timeit
+def transform_gcp_cloud_functions(functions: List[Dict]) -> List[Dict]:
+    """
+    Transforms the raw function data to flatten nested trigger data for ingestion.
+    """
+    transformed_data = []
+    for func_data in functions:
+        func_data['https_trigger_url'] = func_data.get('httpsTrigger', {}).get('url')
+        func_data['event_trigger_type'] = func_data.get('eventTrigger', {}).get('eventType')
+        func_data['event_trigger_resource'] = func_data.get('eventTrigger', {}).get('resource')
+        transformed_data.append(func_data)
+    return transformed_data
+
+
+@timeit
+def load_gcp_cloud_functions(
+    neo4j_session: neo4j.Session,
+    data: List[Dict],
+    project_id: str,
+    update_tag: int,
+) -> None:
+    """
+    Ingests transformed GCP Cloud Functions using the Cartography data model.
+    """
+    logger.info(f"Loading {len(data)} GCP Cloud Functions for project {project_id}.")
+
+    data.sort(key=lambda x: x['name'].split('/')[3] if len(x['name'].split('/')) > 3 else None)
+    for region, functions_in_region in groupby(data, key=lambda x: x['name'].split('/')[3] if len(x['name'].split('/')) > 3 else None):
+        if region:
+            functions_list = list(functions_in_region)
```

Comment:
This looks terrifying lol.

Can we move it to a transform function so that the `load_*()` function is just a dumb wrapper around `load()`, and can we make this look less scary? haha